plugins/module_utils/scaleway.py

```diff
@@ -63,6 +73,101 @@ def parse_pagination_link(header):
+def fetch_state(api, resource):
+    api.module.debug("fetch_state of resource: %s" % resource["id"])
+    response = api.get(path=api.api_path + "/%s" % resource["id"])
+
+    if response.status_code == 404:
+        return "absent"
+
+    if not response.ok:
+        msg = 'Error during state fetching: (%s) %s' % (response.status_code, response.json)
+        api.module.fail_json(msg=msg)
+
+    try:
+        api.module.debug("Resource %s in state: %s" % (resource["id"], response.json["status"]))
+        return response.json["status"]
+    except KeyError:
+        api.module.fail_json(msg="Could not fetch state in %s" % response.json)
+
+
+def fetch_paginated_resources(api, resource_key, **pagination_kwargs):
+    response = api.get(
+        path=api.api_path,
+        params=pagination_kwargs)
+
+    status_code = response.status_code
+    if not response.ok:
+        api.module.fail_json(msg='Error getting {0} [{1}: {2}]'.format(
+            resource_key,
+            response.status_code, response.json['message']))
+
+    return response.json[resource_key]
+
+
+def fetch_all_resources(api, resource_key, **pagination_kwargs):
+    resources = []
+
+    result = [None]
+    while len(result) != 0:
+        result = fetch_paginated_resources(api, resource_key, **pagination_kwargs)
+        resources += result
+        if 'page' in pagination_kwargs:
+            pagination_kwargs['page'] += 1
+        else:
+            pagination_kwargs['page'] = 2
+
+    return resources
+
+
+def wait_to_complete_state_transition(api, resource, stable_states, force_wait=False):
+    wait = api.module.params["wait"]
+
+    if not (wait or force_wait):
+        return
+
+    wait_timeout = api.module.params["wait_timeout"]
+    wait_sleep_time = api.module.params["wait_sleep_time"]
+
+    # Prevent requesting the ressource status too soon
+    time.sleep(wait_sleep_time)
+
+    start = datetime.datetime.utcnow()
+    end = start + datetime.timedelta(seconds=wait_timeout)
+
+    while datetime.datetime.utcnow() < end:
+        api.module.debug("We are going to wait for the resource to finish its transition")
+
+        state = fetch_state(api, resource)
+        if state in stable_states:
+            api.module.debug("It seems that the resource is not in transition anymore.")
+            api.module.debug("load-balancer in state: %s" % fetch_state(api, resource))
+            break
+
+        time.sleep(wait_sleep_time)
+    else:
+        api.module.fail_json(msg="Server takes too long to finish its transition")
+
+
+def filter_sensitive_attributes(container, attributes):
+    for attr in attributes:
+        container[attr] = "SENSITIVE_VALUE"
+
+    return container
+
+
+def resource_attributes_should_be_changed(target, wished, verifiable_mutable_attributes, mutable_attributes):
+    diff = dict()
+    for attr in verifiable_mutable_attributes:
+        if wished[attr] is not None and target[attr] != wished[attr]:
+            diff[attr] = wished[attr]
```

Comment:
You might want to consider using `recursive_diff()` for this, although not sure you need a *deep* diff.

https://github.com/ansible/ansible/blob/HEAD/lib/ansible/module_utils/common/dict_transformations.py#L127-L154

It will import with:

```python
from ansible.module_utils.common.dict_transformations import recursive_diff
```

Reply:
There is a little more logic than just doing a simple `dict` diff.

With this method, we compare the actual state of the resource retreived from API with the desired state described in the Ansible task.
Some attributes of thoses resources doesn't have the same schema on `POST` or `GET` requests.

For example, in a futur module `scaleway_container_namespace` one of the parameter store secret values. When creating the resource, the value is provided as plain text but when querying the resource, the value contains an encoded value.
If we make as simple diff, Ansible is going to find a diff each time the module is called even if the resource doesn't actually require an update.

This is why we have to parameters to this function : 
- `verifiable_mutable_attributes` : List of attributes that we can check if they need to be updated (in my previous example, the secret parameters is not in this list)
- `mutable_attributes` : List of attributes that still should be updated regardless of if we can check if they needed to be changed or not.

Finally in the edge case where only on resource parameter value should be changed and it's on that is not in `verifiable_mutable_attributes` then no update will be triggered. This could be an issue but I'm not sure how to handle it

Reply:
Fair enough